src/metabase/lib/native.cljc

```diff
@@ -15,25 +15,93 @@
    [metabase.util.malli :as mu]
    [metabase.util.malli.registry :as mr]))
 
+(def separators
+  "The tokens used for template tags."
+  #{"{{" "}}" "[[" "]]"})
+
+(defn- next-separator [text start]
+  (let [indexes (map #(or (str/index-of text % start) ##Inf) separators)
+        index   (apply min indexes)]
+    (if (infinite? index)
+      [nil nil]
+      [index (subs text index (+ 2 index))])))
+
+(mu/defn ^:private tokenize-query :- [:sequential :string]
+  "Tokenize the query for easier parsing.
+   This splits the string at the separators defined in `separators`
+   but keeps them in the resulting sequence."
+  [query-text :- ::common/non-blank-string]
+  (let [cnt (count query-text)]
+    (loop [idx 0
+          res []]
+      (if (>= idx cnt)
+        res
+        (let [[jdx tok] (next-separator query-text idx)]
+          (if (nil? jdx)
+            (conj res (subs query-text idx cnt))
+            (recur (+ 2 jdx) (conj res (subs query-text idx jdx) tok))))))))
+
+(mr/def ::template-tag-with-context
+   [:map
+    [:name [:maybe :string]]
+    [:optional :boolean]])
+
+(mu/defn ^:private parse-template-tags :- [:sequential ::template-tag-with-context]
+  "Parse the tokenized query into a sequence of tags, possibly with invalid content."
+  [toks :- [:sequential :string]]
+  (loop [tags []
+         [token & tail] toks
+         in-optional-block false
+         in-template-tag false
+         current-template-text ""]
+    (if (nil? token)
+      tags
+      (case token
+        "{{" (recur tags tail in-optional-block true "")
+        "}}" (if in-template-tag
+              ;; close the current template tag
+              (recur (conj tags {:name current-template-text :optional in-optional-block}) tail in-optional-block false "")
+              (recur tags tail in-optional-block in-template-tag current-template-text))
+        "[[" (recur tags tail (not in-template-tag) in-template-tag current-template-text)
+        "]]" (recur tags tail (if in-template-tag in-optional-block false) in-template-tag current-template-text)
+        (if in-template-tag
+          (recur tags tail in-optional-block in-template-tag (str current-template-text token))
+          (recur tags tail in-optional-block in-template-tag current-template-text))))))
+
 (def ^:private variable-tag-regex
-  #"\{\{\s*([A-Za-z0-9_\.]+)\s*\}\}")
+  #"^([A-Za-z0-9_\.]+)\s*$")
 
 (def ^:private snippet-tag-regex
-  #"\{\{\s*(snippet:\s*[^}]+)\s*\}\}")
+  #"^(snippet:.+)$")
 
 (def ^:private card-tag-regex
-  #"\{\{\s*(#([0-9]*)(-[a-z0-9-]*)?)\s*\}\}")
+  #"^(#[0-9]+(?:-[a-z0-9-]*)?)\s*$")
 
 (def ^:private tag-regexes
-  [variable-tag-regex snippet-tag-regex card-tag-regex])
+  [snippet-tag-regex card-tag-regex variable-tag-regex])
+
+(mu/defn ^:private format-template-tag-name :- [:maybe :string]
+  "Parse and validate a template tag's content."
+  [content :- :string]
+  (second (some #(re-find % (str/triml content)) tag-regexes)))
```

Comment:
The anchors and groups are required to make sure we match the whole string, especially because the whitespace behaviour is different for snippets than it is for other tags (snippets preserve trailing whitespace), ie. the trailing `\s*` groups on `variable-tag-regex` and `card-tag-regex`.

Reply:
Don't we have start and end anchors in all patterns? If we do, then we can remove them and use `re-matches` instead of `re-find`.

Reply:
I'm not sure I understand what you're getting at (probably to me not being super familiar with Clojure).

The anchors are there to avoid matching `variable-tag-regex` to arbitrary (malformed) substrings of the tags.
Eg. `#1o4-foo` should not match anything, but if we remove the anchors, we would match `foo` nonetheless.

The regexp groups are there because in `variable-tag-regex` and `card-tag-regex` we do not want to match trailing whitespace, whereas in `snippet-tag-regex` we do.